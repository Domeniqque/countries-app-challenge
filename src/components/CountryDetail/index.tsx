import React from 'react';
import { ICountry } from '../../models/Country';
import { numberWithDots } from '../../utils/formatNumber';

import {
  Card,
  CardHeader,
  CardBody,
  TopLevelDomain,
} from './styles';

export interface ICountryDetail {
  country?: ICountry;
  loading?: boolean;
}

const CountryDetail: React.FC<ICountryDetail> = ({ country, loading }) => {
  if (!country || loading) {
    return <p>Loading...</p>;
  };

  return (
    <Card>
      <CardHeader>
        <img src={country.flag.svgFile} alt="Country flag" width={120} />

        <div>
          <h1>{country.name}</h1>
          <p>{country.capital}</p>
        </div>
      </CardHeader>

      <CardBody>
        <section>
          {country.area && (
            <div>
              <h6>Area</h6>
              <p>{`${numberWithDots(country.area)} kmÂ²`}</p>
            </div>
          )}

          <div>
            <h6>Population</h6>
            <p>{numberWithDots(country.population)}</p>
          </div>
        </section>

        <TopLevelDomain>
          <details>
            <summary>Top Level Domains</summary>
            <p>
              A country code top-level domain (ccTLD) is an Internet top-level
              domain generally used or reserved for a country, sovereign state,
              or dependent territory identified with a country code.&nbsp;
              <a
                rel="noopener noreferrer"
                target="_blank"
                href="https://en.m.wikipedia.org/wiki/Country_code_top-level_domain"
              >
                Wikipedia
              </a>
            </p>
          </details>

          <ul>
            {country.topLevelDomains?.map(tld => (
                <li key={`tld-${tld._id}`}>
                  <p>
                    <span>www.example</span>
                    <strong>{tld.name}</strong>
                  </p>
                </li>
              ))}
          </ul>
        </TopLevelDomain>
      </CardBody>
    </Card>
  );
};

export default CountryDetail;
