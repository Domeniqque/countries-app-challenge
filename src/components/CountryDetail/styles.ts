import styled from 'styled-components';

export const Card = styled.div`
  margin-top: 20px;
  background: var(--white);
  padding: 24px 16px;
  box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.1);
`;

export const CardHeader = styled.header`
  display: flex;
  flex-direction: row;
  align-items: center;

  > div {
    flex: 1;
    padding-left: 18px;

    h1 {
      color: var(--dark-blue);
      font-size: 1.8rem;
      line-height: 28px;
    }

    p {
      color: var(--grey300);
      font-size: 1rem;
      margin-top: 2px;
    }
  }
`;

export const CardBody = styled.main`
  > section {
    display: flex;
    flex-direction: row;
    justify-content: space-around;
    margin: 40px 0;

    div {
      text-align: center;

      h6 {
        text-transform: uppercase;
        font-size: 0.8rem;
        color: var(--grey400);
        font-weight: 400;
      }

      p {
        font-size: 1.6rem;
        color: var(--light-blue);
        line-height: 28px;
        font-weight: 400;
      }
    }
  }
`;

export const TopLevelDomain = styled.section`
  justify-content: flex-start !important;
  flex-direction: column !important;

  details {
    summary {
      font-size: 1rem;
      font-weight: 700;
      color: var(--dark-blue);
      cursor: pointer;

      &::-webkit-details-marker {
        color: var(--dark-blue);
      }

      &:hover {
        text-decoration: underline;
      }
    }

    p {
      font-size: 0.8rem;
      color: var(--grey400);
      margin-top: 4px;
    }
  }

  ul {
    margin-left: 20px;

    li {
      padding: 4px 0;

      p {
        font-size: 1.2rem;
        font-style: italic;
        color: var(--grey300);

        strong {
          color: var(--light-blue);
        }
      }
    }
  }
`;
