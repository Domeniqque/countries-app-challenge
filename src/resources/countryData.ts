import { ICountry } from '../models/Country';

const COUNTRY_DATA: ICountry = {
  _id: '3',
  name: 'Afghanistan',
  capital: 'Kabul',
  area: 652230,
  population: 27657145,
  location: {
    latitude: 33,
    longitude: 65,
  },
  topLevelDomains: [
    {
      _id: '43',
      name: '.af',
      countries: [
        {
          _id: '4',
          name: 'Afghanistan',
        },
      ],
    },
  ],
  flag: {
    svgFile: 'https://restcountries.eu/data/afg.svg',
    emoji: 'ðŸ‡¦ðŸ‡«',
  },
};

const COUNTRY_DATA_LIST: ICountry[] = [
  {
    _id: '3',
    name: 'Afghanistan',
    capital: 'Kabul',
    flag: {
      emoji: 'ðŸ‡¦ðŸ‡«',
    },
  },
  {
    _id: '27',
    name: 'Ã…land Islands',
    capital: 'Mariehamn',
    flag: {
      emoji: 'ðŸ‡¦ðŸ‡½',
    },
  },
  {
    _id: '661',
    name: 'Brazil',
    capital: 'BrasÃ­lia',
    flag: {
      emoji: 'ðŸ‡§ðŸ‡·',
    },
  },
];

export { COUNTRY_DATA, COUNTRY_DATA_LIST };
