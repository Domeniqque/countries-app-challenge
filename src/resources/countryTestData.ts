import { ICountry } from '../models/Country';
import { INearCountry } from '../services/api/country';

const COUNTRY_DATA_LIST: ICountry[] = [
  {
    _id: '561',
    name: 'Bolivia (Plurinational State of)',
    capital: 'Sucre',
    area: '1098581',
    population: '10985059',
    location: {
      latitude: -17,
      longitude: -65,
    },
    topLevelDomains: [
      {
        _id: '562',
        name: '.bo',
      },
    ],
    flag: {
      svgFile: 'https://restcountries.eu/data/bol.svg',
      emoji: 'ðŸ‡§ðŸ‡´',
    },
  },
  {
    _id: '3278',
    name: 'Paraguay',
    capital: 'AsunciÃ³n',
    area: '406752',
    population: '6854536',
    location: {
      latitude: -23,
      longitude: -58,
    },
    topLevelDomains: [
      {
        _id: '3279',
        name: '.py',
      },
    ],
    flag: {
      svgFile: 'https://restcountries.eu/data/pry.svg',
      emoji: 'ðŸ‡µðŸ‡¾',
    },
  },
  {
    _id: '661',
    name: 'Brazil',
    capital: 'BrasÃ­lia',
    area: '8515767',
    population: '206135893',
    location: {
      latitude: -10,
      longitude: -55,
    },
    topLevelDomains: [
      {
        _id: '662',
        name: '.br',
      },
    ],
    flag: {
      svgFile: 'https://restcountries.eu/data/bra.svg',
      emoji: 'ðŸ‡§ðŸ‡·',
    },
  },
];

const BOLIVIA_DATA = COUNTRY_DATA_LIST[0];

const BRAZIL_DATA = COUNTRY_DATA_LIST[2];

const BRAZIL_COORDINATES: INearCountry = {
  countryName: 'Brazil',
  distanceInKm: 1481.9677422904354,
};

export { BOLIVIA_DATA, COUNTRY_DATA_LIST, BRAZIL_COORDINATES, BRAZIL_DATA };
