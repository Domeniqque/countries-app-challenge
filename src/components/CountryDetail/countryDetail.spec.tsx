import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

import CountryDetail, { ICountryDetail } from './index';
import { COUNTRY_DATA } from '../../resources/countryData';

const setup = (param?: ICountryDetail) => {
  // eslint-disable-next-line react/jsx-props-no-spreading
  const utils = render(<CountryDetail {...param} />, { wrapper: MemoryRouter });

  return { ...utils };
};

describe('CountryDetail', () => {
  it('should render loading', () => {
    const { getByText } = setup({ loading: true, country: undefined });

    expect(getByText(/loading/i)).toBeInTheDocument();
  });

  it('should render country detail', () => {
    const country = COUNTRY_DATA;
    const { queryByText, getByRole, getByText } = setup({
      loading: false,
      country,
    });

    expect(queryByText(/loading/i)).toBeNull();
    expect(getByRole(/img/).getAttribute('src')).toBe(country.flag.svgFile);
    expect(getByText(country.name)).toBeInTheDocument();
  });

  it('render correctly details', () => {
    const component = renderer
      .create(<CountryDetail country={COUNTRY_DATA} />)
      .toJSON();

    expect(component).toMatchSnapshot();
  });
});
