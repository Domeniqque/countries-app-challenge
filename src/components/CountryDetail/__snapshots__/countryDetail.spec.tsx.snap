// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryDetail render correctly details 1`] = `
<div
  className="sc-bdnylx hMvujY"
>
  <header
    className="sc-gtssRu dTqkMi"
  >
    <img
      alt="Country flag"
      src="https://restcountries.eu/data/afg.svg"
      width={120}
    />
    <div>
      <h1>
        Afghanistan
      </h1>
      <p>
        Kabul
      </p>
    </div>
  </header>
  <main
    className="sc-dlnjPT fqZdor"
  >
    <section>
      <div>
        <h6>
          Area
        </h6>
        <p>
          652.230 km²
        </p>
      </div>
      <div>
        <h6>
          Population
        </h6>
        <p>
          27.657.145
        </p>
      </div>
    </section>
    <section
      className="sc-hKFyIo ehJLqX"
    >
      <details>
        <summary>
          Top Level Domains
        </summary>
        <p>
          A country code top-level domain (ccTLD) is an Internet top-level domain generally used or reserved for a country, sovereign state, or dependent territory identified with a country code. 
          <a
            href="https://en.m.wikipedia.org/wiki/Country_code_top-level_domain"
            rel="noopener noreferrer"
            target="_blank"
          >
            Wikipedia
          </a>
        </p>
      </details>
      <ul>
        <li>
          <p>
            <span>
              www.example
            </span>
            <strong>
              .af
            </strong>
          </p>
        </li>
      </ul>
    </section>
  </main>
</div>
`;
